### PR TITLE
Analytics opt in for posthog

### PR DESCRIPTION
**Pull request by novocaine**
Design: https://github.com/vector-im/internal-planning/issues/157

See https://github.com/vector-im/element-meta/issues/10 for the epic.

We have added an analytics token to be able to understand people's usage across different devices, so need to re-prompt users for their consent to this new scheme.

* Add a new flag `pseudonymousAnalyticsOptIn` replacing `analyticsOptIn`, stored at account level, so people only need to opt in once.
* Show this toast in login to users that have `analyticsOptIn` set but not yet `pseudonymousAnalyticsOptIn`
Screenshot 2021-11-30 at 10 59 40

* Show this prompt to new users with no analytics opt-in set at all

Screenshot 2021-11-30 at 10 58 39

* New preferences wording

Screenshot 2021-11-19 at 11 16 35

* New "learn more" dialog

Screenshot 2021-11-30 at 11 00 44

* Don't notify users that previously opted out.

* Support a new config flag `analyticsOwner` which is used in these toasts when explaining which entity the data is sent to ("Help improve %(analyticsOwner)"). If unset, display `brand`. This allows deployments whose brand differs from the receiver of the analytics to explain the situation to their users (e.g. AcmeCorp badges their app, but explains the data is sent to Element, not them)

* The new opt-in and flags are only used when posthog is configured; prior to that there are no changes to UX or tracking behaviour.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Analytics opt in for posthog ([\#6936](https://github.com/matrix-org/matrix-react-sdk/pull/6936)). Contributed by @novocaine.

Preview: https://61ad3b44e27107a7054bee43--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by novocaine** [comment]
> Where does the link in "analytics data" redirect to? 

It pops up this (already existed, you can also see it by going to settings -> security and privacy -> privacy -> "Learn more about how we use analytics")

Screenshot 2021-10-17 at 08 59 30

This is kind of neat, but its a bit incomplete / prone to going out of date and I agree we should link to the cookie policy .. instead?

**Comment by novocaine** [comment]
@deniseal can you review the new copy from design please? I've updated the screenshots with the latest

**Comment by novocaine** [comment]
To test this in the netlify build for the new user flow, just sign in

Test test the flow for existing opted-in users, sign in to netlify then open the console and do this

```
mxSettingsStore.setValue("analyticsOptIn", null, "device", true)
mxSettingsStore.setValue("pseudonymousAnalyticsOptIn", null, "account", null)
```
After a moment, you should see the toast.

**Review by nadonomy — CHANGES_REQUESTED** [review]
@novocaine this is looking/feeling great! Couple of q's/notes:

Screenshot 2021-11-24 at 11 00 05

The design looked a bit visually imbalanced, so I tweaked a few values to get the screenshot above:

- `mx_AnalyticsLearnMore_image_holder`: [Echoing the 20px spacing on both the top & bottom](https://user-images.githubusercontent.com/4626865/143226532-faf01c7c-488f-4d8e-b71e-39f0763e161d.png) feels better (I used margin).
- `mx_Dialog_content`: The 68px bottom margin is a bit abstract. I realise this is applied globally so might be a risky change, but the modal looks a lot better without it. Is there a simple tweak we could do to this modal specifically?

Otherwise, looking great! Did you deliberately omit the link to read all terms?

File: res/css/views/dialogs/_AnalyticsLearnMoreDialog.scss [hunk 1]
```diff
@@ -0,0 +1,43 @@
+.mx_AnalyticsLearnMoreDialog {
```
Comment (t3chguy) [inline res/css/views/dialogs/_AnalyticsLearnMoreDialog.scss:1]: copyright missing

File: src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+const AnalyticsLearnMoreDialog: React.FC<IProps> = (
+    {
+        onFinished,
+        analyticsOwner,
+        privacyPolicyUrl,
+        primaryButton,
+        cancelButton,
+        hasCancel,
+    }) => {
```
Comment (t3chguy) [inline src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx:46]: this indentation feels awkward, normally we'd keep the opening brace on L38 and then un-indent L39-L46

File: src/toasts/AnalyticsToast.tsx [hunk 3]
```diff
@@ -23,16 +23,51 @@ import Analytics from "../Analytics";
...
 import ToastStore from "../stores/ToastStore";
+import {
+    ButtonClicked,
+    showDialog as showAnalyticsLearnMoreDialog,
+} from "../components/views/dialogs/AnalyticsLearnMoreDialog";
 
 const onAccept = () => {
     dis.dispatch({
-        action: 'accept_cookies',
+        action: 'pseudonymous_analytics_accept',
```
Comment (t3chguy) [inline src/toasts/AnalyticsToast.tsx:33]: can this please be added to the Action enum?

File: src/toasts/AnalyticsToast.tsx [hunk 4]
```diff
@@ -42,37 +77,82 @@ const onUsageDataClicked = () => {
...
+    }
+    showToast(props);
+};
+
+export const showAnonymousAnalyticsOptInToast = (): void => {
+    const props = {
+        description: getAnonymousDescription(),
+        acceptLabel: _t("Yes"),
+        onAccept: () => dis.dispatch({
+            action: 'anonymous_analytics_accept',
```
Comment (t3chguy) [inline src/toasts/AnalyticsToast.tsx:148]: Can this please be added to the Action enum?

File: src/toasts/AnalyticsToast.tsx [hunk 5]
```diff
@@ -42,37 +77,82 @@ const onUsageDataClicked = () => {
...
+        }),
+        rejectLabel: _t("No"),
+        onReject: () => dis.dispatch({
+            action: "anonymous_analytics_reject",
```
Comment (t3chguy) [inline src/toasts/AnalyticsToast.tsx:152]: Can this please be added to the Action enum?

**Review by t3chguy — CHANGES_REQUESTED** [review]

**Comment by novocaine** [comment]
> @novocaine this is looking/feeling great! Couple of q's/notes:
> 
> Screenshot 2021-11-24 at 11 00 05
> 
> The design looked a bit visually imbalanced, so I tweaked a few values to get the screenshot above:
> 
> * `mx_AnalyticsLearnMore_image_holder`: [Echoing the 20px spacing on both the top & bottom](https://user-images.githubusercontent.com/4626865/143226532-faf01c7c-488f-4d8e-b71e-39f0763e161d.png) feels better (I used margin).
> * `mx_Dialog_content`: The 68px bottom margin is a bit abstract. I realise this is applied globally so might be a risky change, but the modal looks a lot better without it. Is there a simple tweak we could do to this modal specifically?
> 

Done

> Otherwise, looking great! Did you deliberately omit the link to read all terms?

oops, this got lost after a refactor

File: src/Analytics.tsx [hunk 6]
```diff
@@ -393,11 +380,18 @@ export class Analytics {
...
             title: _t('Analytics'),
             description: <div className="mx_AnalyticsModal">
-                <div>{ _t('The information being sent to us to help make %(brand)s better includes:', {
+                { cookiePolicyUrl &&
+                <p>
+                    Our complete cookie policy can be found
+                    <a href={cookiePolicyUrl} target="_blank" rel="noreferrer noopener">here</a>.
+                </p>
+                }
+                <div>{ _t('Some examples of the information being sent to us to help make %(brand)s better includes:', {
```
[the remark this answers is not in the document]
Comment (novocaine) [reply src/Analytics.tsx:410]: We are recording the data dictionary over [here](https://github.com/matrix-org/matrix-analytics-events) in JSON schema. Clients cant send events without adding to this schema.

So, we can generate something human readable from this. I'll update you once that's complee.

File: src/settings/Settings.tsx [hunk 7]
```diff
@@ -239,13 +238,6 @@ export const SETTINGS: {[setting: string]: ISetting} = {
         supportedLevels: LEVELS_FEATURE,
         default: false,
     },
-    "feature_pseudonymous_analytics_opt_in": {
```
[the remark this answers is not in the document]
Comment (novocaine) [reply src/settings/Settings.tsx:304]: https://github.com/vector-im/element-web/pull/19900

**Comment by deniseal** [comment]
@novocaine a few notes:
- I would recommend changing the colours of the buttons of the various toasts. The ICO's guidance tends to be that both options need to be of equal footing, for consent to be considered fair. They specifically call out a red vs green juxtaposition as not appropriate.
- technically, a randomly generated identifier is still personal data (albeit indirect data). In the last 'learn more' dialogue this might be misleading to end users, so I would change "personal data" to "account data". 
- the last 'learn more' dialogue should have a link to the full cookie policy.

**Review by t3chguy — APPROVED** [review]

**Comment by nadonomy** [comment]
> * I would recommend changing the colours of the buttons of the various toasts. The ICO's guidance tends to be that both options need to be of equal footing, for consent to be considered fair. They specifically call out a red vs green juxtaposition as not appropriate.

@novocaine the red vs green also isn't desirable for UX. To amend, please can we:

1. Use solid CTAs on equal footing for the analytics toasts (updated [Figma](https://www.figma.com/file/nt8rhrQ9arytjKUl5omfYj/%23157-Analytics-prompt?node-id=50%3A12568) here) per the ICOs guidance.
2. Use hollow/ghost style buttons for the secondary actions in the other FTUE toasts.

**Comment by novocaine** [comment]
> > * I would recommend changing the colours of the buttons of the various toasts. The ICO's guidance tends to be that both options need to be of equal footing, for consent to be considered fair. They specifically call out a red vs green juxtaposition as not appropriate.
> 
> @novocaine the red vs green also isn't desirable for UX. To amend, please can we:
> 
> 1. Use solid CTAs on equal footing for the analytics toasts (updated [Figma](https://www.figma.com/file/nt8rhrQ9arytjKUl5omfYj/%23157-Analytics-prompt?node-id=50%3A12568) here) per the ICOs guidance.

done

> 2. Use hollow/ghost style buttons for the secondary actions in the other FTUE toasts.

Not sure which other toasts actually have a red button atm?

> 
> > technically, a randomly generated identifier is still personal data (albeit indirect data). In the last 'learn more' dialogue this might be misleading to end users, so I would change "personal data" to "account data".
> 
> Amended in Figma.

done

File: src/PosthogAnalytics.ts [hunk 8]
```diff
@@ -132,7 +132,7 @@ export class PosthogAnalytics {
 
     private anonymity = Anonymity.Disabled;
     // set true during the constructor if posthog config is present, otherwise false
-    private enabled = false;
+    private readonly enabled: boolean = false;
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/PosthogAnalytics.ts:135]: can we get a comment to say that it's decided in the constructor? Our typical pattern isn't to create new instances as needed, but rather toggle an internal flag.

File: src/PosthogAnalytics.ts [hunk 9]
```diff
@@ -351,12 +329,31 @@ export class PosthogAnalytics {
...
 
-    public async updateAnonymityFromSettings(userId?: string): Promise<void> {
+    public async updateAnonymityFromSettings(pseudonymousOptIn: boolean): Promise<void> {
         // Update this.anonymity based on the user's analytics opt-in settings
-        // Identify the user (via hashed user ID) to posthog if anonymity is pseudonmyous
-        this.setAnonymity(PosthogAnalytics.getAnonymityFromSettings());
-        if (userId && this.getAnonymity() == Anonymity.Pseudonymous) {
+        const anonymity = pseudonymousOptIn ? Anonymity.Pseudonymous : Anonymity.Disabled;
+        this.setAnonymity(anonymity);
+        if (anonymity == Anonymity.Pseudonymous) {
```
Comment (turt2live) [inline src/PosthogAnalytics.ts:336]:
```suggestion
        if (anonymity === Anonymity.Pseudonymous) {
```

File: src/PosthogAnalytics.ts [hunk 10]
```diff
@@ -351,12 +329,31 @@ export class PosthogAnalytics {
...
             await this.identifyUser(MatrixClientPeg.get(), PosthogAnalytics.getRandomAnalyticsId);
         }
+
+        if (anonymity != Anonymity.Disabled) {
```
Comment (turt2live) [inline src/PosthogAnalytics.ts:340]:
```suggestion
        if (anonymity !== Anonymity.Disabled) {
```

File: src/dispatcher/actions.ts [hunk 11]
```diff
@@ -215,4 +215,25 @@ export enum Action {
...
+    /**
+     * The user accepted anonymous analytics (i.e. matomo, pre-posthog) from the toast
+     * (this action and its handler can be removed once posthog is rolled out)
+     */
+    AnonymousAnalyticsAccept = "anonymous_analytics_accept",
+
+    /**
+     * The user rejected anonymous analytics (i.e. matomo, pre-posthog) from the toast
+     */
+    AnonymousAnalyticsReject = "anonymous_analytics_reject"
```
Comment (turt2live) [inline src/dispatcher/actions.ts:238]: We've been really bad about it lately, but can we get payload information included in the comments on these? Specifically, an explicit "no additional payload required" or "should be used with SuperCoolPayload" please.

File: src/toasts/AnalyticsToast.tsx [hunk 12]
```diff
@@ -23,16 +23,52 @@ import Analytics from "../Analytics";
...
+        primaryButton: _t("Enable"),
+    });
+};
+
+const onLearnMorePreviouslyOptedIn = () => {
+    showAnalyticsLearnMoreDialog({
+        onFinished: (buttonClicked?: ButtonClicked) => {
+            if (buttonClicked === ButtonClicked.Primary) {
+                // user clicked "That's fine"
+                onAccept();
```
Comment (turt2live) [inline src/toasts/AnalyticsToast.tsx:63]: I might be blind, but where does `onAccept` come from?

File: src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+interface IProps {
+    onFinished?(buttonClicked?: ButtonClicked): void;
+    analyticsOwner: string;
+    privacyPolicyUrl?: string;
+    primaryButton?: string;
+    cancelButton?: string;
+    hasCancel?: boolean;
+}
+
+const AnalyticsLearnMoreDialog: React.FC<IProps> = ({
```
Comment (turt2live) [inline src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx:38]: This dialog looks eerily similar to the new Compound dialog from design. While this PR has been open, a new base class `ScrollableBaseModal` was introduced which might fit the design (at least, it does from what I can see at a glance).

I believe we're supposed to be using Compound for everything these days, so it'd be good to either align the dialog with that or fix the Compound design to match the proposal here.

**Review by turt2live — COMMENTED** [review]
overall the changes look fine, just a few new questions/comments given PR age (sorry).

File: src/toasts/AnalyticsToast.tsx (hunk 12 above)
Comment (novocaine) [reply src/toasts/AnalyticsToast.tsx:63]: https://github.com/matrix-org/matrix-react-sdk/pull/6936/files/560494ff65cbc63edd590496c57da850e20099e6#diff-4386d47cc2349bae17a0f01e44b162d774569285c3987d6374334fd9f0bff3ccR32

File: src/PosthogAnalytics.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (novocaine) [reply src/PosthogAnalytics.ts:135]: it does have an existing comment, happy to reword it if you have a suggestion that'd make it 
 clearer

**Review by deniseal — APPROVED** [review]
All looks good from a privacy perspective following the latest changes.

File: src/toasts/AnalyticsToast.tsx (hunk 12 above)
Comment (turt2live) [reply src/toasts/AnalyticsToast.tsx:63]: ah ha, thanks!

File: src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx (hunk 13 above)
Comment (novocaine) [reply src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx:38]: @nadonomy - should this dialog be using Compound?

@turt2live - Where do I go to view the dialog in Compound which you're referring to?

Comment (turt2live) [reply src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx:38]: I never know where to find it, but polls use a new kind of dialog that is supposed to (eventually) be used everywhere, aiui.

Comment (nadonomy) [reply src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx:38]: > @nadonomy - should this dialog be using Compound?

We haven't set a formal mandate for new feature development to use new components yet.

If it's trivial to swap out now with the class Travis mentioned then sgtm and we save some debt/cleanup later, but wouldn't block on it.

Comment (novocaine) [reply src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx:38]: It's not a huge amount of work but non-trivial - also I think Travis is suggesting the new dialog looks slightly different..?

Comment (nadonomy) [reply src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx:38]: There might be a subtle visual difference, comparing the normal to scrollable versions here:

https://www.figma.com/file/X4XTH9iS2KGJ2wFKDqkyed/Compound?node-id=643%3A101

But I think we should just move forward on this PR:

1. This modal isn't scrollable.
2. I think we actually just want one one true modal component which _can_ scroll when needed.
3. In future we plan to standardise on re-usable components. If this component isn't documented well enough (Storybook? Document somewhere detailing components and use? Better notes in Figma? etc) I think we're worrying prematurely.

So would just prefer to stick the course on analytics and avoid this as it's not immediately impactful and as you said non-trivial.

Comment (turt2live) [reply src/components/views/dialogs/AnalyticsLearnMoreDialog.tsx:38]: hmm, as you say: the dialog is different enough to be, well, different. Probably not worth worrying about here, I guess.

It is more debt we have to worry about in the future, but it doesn't sound like there's a firm requirement to use the new dialog structure yet, so would rather steer away from accidentally implementing a 5th dialog style that may or may not get used.

**Review by turt2live — APPROVED** [review]
thanks!

**Review by nadonomy — APPROVED** [review]
LGTM.

File: res/css/views/toasts/_AnalyticsToast.scss [hunk 14]
```diff
@@ -15,13 +15,17 @@ limitations under the License.
...
+        border: 1px solid $accent;
+        font-weight: 600;
     }
 
     .mx_AccessibleButton_kind_primary {
-        background: $accent;
+        background-color: $accent;
         color: #ffffff;
+        border: 1px solid $accent;
+        font-weight: 600;
```
Comment (t3chguy) [inline res/css/views/toasts/_AnalyticsToast.scss:29]: we have $font-semi-bold for this

**Review by t3chguy — APPROVED** [review]

**Pull request merged**